src/serde/ser.rs

```diff
@@ -0,0 +1,804 @@
+    where
+        T: Serialize,
+    {
+        value.serialize(&mut **self)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.writer.step_out()
+    }
+}
+
+impl<'a, E> ser::SerializeTuple for &'a mut Serializer<E>
+where
+    E: Writer,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    fn serialize_element<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        value.serialize(&mut **self)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.writer.step_out()
+    }
+}
+
+impl<'a, E> ser::SerializeTupleStruct for &'a mut Serializer<E>
+where
+    E: Writer,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    fn serialize_field<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        value.serialize(&mut **self)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.writer.step_out()
+    }
+}
+
+impl<'a, E> ser::SerializeTupleVariant for &'a mut Serializer<E>
+where
+    E: Writer,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    fn serialize_field<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        value.serialize(&mut **self)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.writer.step_out()
+    }
+}
+
+impl<'a, E> ser::SerializeMap for &'a mut Serializer<E>
+where
+    E: Writer,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    fn serialize_key<T: ?Sized>(&mut self, key: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        // We need to verify that the key is a string type or can be converted
+        // to string
+        let mk_serializer = MapKeySerializer {};
+        let field: String = key.serialize(mk_serializer)?;
+        Ok(self.writer.set_field_name(field))
+    }
+
+    fn serialize_value<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        value.serialize(&mut **self)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.writer.step_out()
+    }
+}
+
+impl<'a, E> ser::SerializeStructVariant for &'a mut Serializer<E>
+where
+    E: Writer,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    fn serialize_field<T: ?Sized>(
+        &mut self,
+        key: &'static str,
+        value: &T,
+    ) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.writer.set_field_name(key);
+        value.serialize(&mut **self)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.writer.step_out()
+    }
+}
+
+struct MapKeySerializer {}
+
+fn key_must_be_a_string() -> IonError {
+    IonError::EncodingError {
+        description: "ion does not support non-string keys for maps".to_string(),
+    }
+}
+
+impl<'a> ser::Serializer for MapKeySerializer {
```

Comment:
This is a clever solution! However, I wonder if applications that rely on honest-to-goodness map types would prefer to retain the key's data types.

Some options:
1. Simply err if the data cannot be represented sanely using an Ion `struct`. (that is: ONLY text keys)
2. Convert all keys to strings on a best-effort basis to produce a `struct`. (This is lossy.)
3. Serialize maps as lists of key/value pairs instead of `struct`s.

For the moment I think I'm inclined to go with the first option until we can decide on a final strategy. If we choose either 2 or 3, we'll be inventing a serialization format that's not-quite-Ion. Other Ion implementations won't recognize them the way that this serde implementation would.